src/ddmd/link.d

```diff
@@ -1039,4 +983,384 @@ version (Windows)
+                else if(x64 && FileName.exists(sdk / arch / "kernel32.lib")) // SDK 7.1 or earlier
+                    return sdk / arch;
+                else if(!x64 && FileName.exists(sdk / "kernel32.lib")) // SDK 7.1 or earlier
+                    return sdk;
+            }
+            return null;
+        }
+
+        // iterate through subdirectories named by SDK version in baseDir and return the
+        //  one with the largest version that also contains the test file
+        static const(char)* findLatestSDKDir(const(char)* baseDir, const(char)* testfile)
+        {
+            auto allfiles = FileName.combine(baseDir, "*");
+            core.sys.windows.winbase.WIN32_FIND_DATAA fileinfo;
+            HANDLE h = FindFirstFileA(allfiles, &fileinfo);
+            if (h == INVALID_HANDLE_VALUE)
+                return null;
+
+            const(char)* res = null;
+            do
+            {
+                if (fileinfo.cFileName[0] >= '1' && fileinfo.cFileName[0] <= '9')
+                    if (res is null || strcmp(res, fileinfo.cFileName.ptr) < 0)
+                        if (FileName.exists(FileName.combine(FileName.combine(baseDir, fileinfo.cFileName.ptr), testfile)))
+                            res = strdup(fileinfo.cFileName.ptr);
+            }
+            while(FindNextFileA(h, &fileinfo));
+            FindClose(h);
+
+            return res;
+        }
+
+    }
+
+    struct PathBuilder
+    {
+        const(char)* p;
+        alias p this;
+
+        PathBuilder opDiv(const(char)* name)
+        {
+            if (!name)
+                return this;
+            return PathBuilder(FileName.combine(p, name));
+        }
+    }
+
+    struct RegKey
+    {
+        pragma(lib, "advapi32.lib");
+
+        this(HKEY root, const(char)* szRegPath, bool x64hive = false)
+        {
+            version(Win64)
+            {
+                if (!x64hive)
+                {
+                    size_t len = strlen(szRegPath);
+                    assert(len > 9 && szRegPath[0..9] == r"SOFTWARE\");
+                    char* npath = cast(char*)mem.xmalloc(len + 13);
+                    strcpy(npath, r"SOFTWARE\WOW6432Node\");
+                    strcpy(npath + 21, szRegPath + 9);
+                    szRegPath = npath;
+                }
+            }
+            enum KEY_WOW64_64KEY = 0x000100; // not defined in core.sys.windows.winnt due to restrictive version
+            enum KEY_WOW64_32KEY = 0x000200;
+            LONG lRes = RegOpenKeyExA(root, szRegPath, (x64hive ? KEY_WOW64_64KEY : KEY_WOW64_32KEY), KEY_READ, &key);
+            if (FAILED(lRes))
+                key = null;
+        }
+
+        ~this()
+        {
+            Close();
+        }
+
+        void Close()
+        {
+            if(key)
+            {
+                RegCloseKey(key);
+                key = null;
+            }
+        }
+
+        void Create(HKEY root, wstring keyname, bool x64hive = false)
```

Comment:
I generally dislike using default arguments unless doing it for backwards compatibility. Since this is new code, please use explicit arguments.